rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,578 @@
+	// rel_i starts at zero.
+	let rel_i : u16 = (i - old_block_count as u32) as u16;
+
+	// (to do -- handle case where allocating this way has run out.)
+	let block_id : u16 = (old_total_blocks + rel_i as u64) as u16;
+
+	println!(50, "  region_grow id={} (slot index) i={} block_id={}", (*r).id, i, block_id);
+
+	// Update stable memory with new association.
+	let assoc =  Some((RegionId::from_id((*r).id), i as u16));
+	meta_data::block_region_table::set(BlockId(block_id), assoc.clone());
+
+	if true { // temp sanity testing: read back the data we just wrote.
+	    let assoc_ = meta_data::block_region_table::get(BlockId(block_id));
+	    assert_eq!(assoc, assoc_)
+	}
+
+	new_pages.set_ith_block_id(i, &BlockId(block_id));
+
+	if true { // temp sanity testing: read back the data we just wrote.
+	    let block_id_ = new_pages.get_ith_block_id(i);
+	    assert_eq!(BlockId(block_id), block_id_);
+	}
+    }
+
+    (*r).vec_pages = new_vec_pages;
+    old_page_count.into()
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_load_byte<M: Memory>(_mem: &mut M, r: Value, offset: u64) -> u32 {
+    let r = RegionObject::from_value(&r);
+    let abs_off = r.relative_into_absolute_offset(offset);
+    let mut dst : [u8; 1] = [0];
+    crate::ic0_stable::nicer::read(abs_off, &mut dst);
+    println!(80, "region_load_byte({:?}, {} ~> {}) ==> {}", r.id(), offset, abs_off, dst[0]);
+    dst[0] as u32
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_store_byte<M: Memory>(_mem: &mut M, r: Value, offset: u64, byte: u32) {
+    let r = RegionObject::from_value(&r);
+    let abs_off = r.relative_into_absolute_offset(offset);
+    let src : [u8; 1] = [byte as u8];
+    println!(80, "region_store_byte({:?}, {} ~> {}, {})", r.id(), offset, abs_off, byte);
+    crate::ic0_stable::nicer::write(abs_off, &src);
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_load_blob<M: Memory>(_mem: &mut M, _r: Value, _start: Value, _len: Value) -> Value {
+    rts_trap_with("TODO region_load_blob");
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_store_blob<M: Memory>(_mem: &mut M, _r: Value, _start: Value, _v: Value) {
+    /*
+    let b : *const Blob = v.as_blob();
+    let l = b.len();
+    let p = b.payload_addr();
+    let r : RegionObject = RegionObject(r.as_region());
+    r.relative_into_absolute(
+    */
+    rts_trap_with("TODO region_store_blob");
+}
+
+
+#[ic_mem_fn]
+pub unsafe fn region_next_id<M: Memory>(_mem: &mut M) -> Value {
+    Value::from_scalar(meta_data::total_allocated_regions::get() as u32)
+}
+
+#[ic_mem_fn]
+pub unsafe fn region0_size<M: Memory>(mem: &mut M) -> u64 {
+    region_size(mem, Value::from_ptr(crate::memory::ic::REGION_0 as usize))
+}
+
+#[ic_mem_fn]
+pub unsafe fn region0_grow<M: Memory>(mem: &mut M, new_pages: u64) -> u64 {
+    region_grow(mem, Value::from_ptr(crate::memory::ic::REGION_0 as usize), new_pages)
+}
+
+unsafe fn region0_load<M: Memory>(_mem: &mut M, offset:u64, dst: &mut [u8]) {
+    let r = RegionObject(crate::memory::ic::REGION_0);
```

Comment:
This code won't work correctly if the blocks aren't contiguous, right? In general, you will need to do several low-level reads from different blocks, especially if dst is large, but even when dst is small but you are near the end of a block.

Ditto for region0_store.

Reply:
Indeed!  Thanks for noting that here.  I was reusing the (simple) logic for loading and storing bytes, but hadn't generalized it enough yet.  I had almost forgotten about the need to generalize it.

Hopefully that will be simple to do, and it can all be in one place by generalizing this overly-simple version.   In the recent commits, I went forward trying to use this slice-based API to do the other operations, and it was pretty easy.

Fortunately, the more general logic is not actually not needed until there are multiple regions since the current logic will allocate the single region0 (or any other single-test region) into contiguous blocks.

Reply:
Makes sense. Btw, once done properly, I think it might be enough to only do the bounds check on the high address.